back/engines/commercial/analytics/app/controllers/analytics/web_api/v1/analytics_controller.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'query'
+
+module Analytics
+  module WebApi::V1
+    class AnalyticsController < ::ApplicationController
+      def create
+        authorize :analytics, policy_class: AnalyticsPolicy
+
+        results, errors, response_status = if params[:query].instance_of?(Array)
+          handle_multiple(params[:query])
+        else
+          handle_single(params[:query])
+        end
+
+        if [{}, []].include? errors
+          render json: { 'data' => results }
+        else
+          render json: { 'messages' => errors }, status: response_status
+        end
+      end
+
+      private
+
+      def handle_single(json_query)
+        query = Query.new(json_query)
+        query.validate
+
+        query.run if query.valid
+
+        results = query.results
+        errors = query.error_messages
+        response_status = query.response_status
+
+        [results, errors, response_status]
+      end
+
+      def handle_multiple(json_queries)
+        results = []
+        errors = {}
+        statuses = []
+        queries = []
+
+        json_queries.each_with_index do |json_query, index|
+          query = Query.new(json_query)
+          queries.push(query)
+          query.validate
+          next if query.valid
+
+          errors[index] = query.error_messages
+          statuses.push(query.response_status)
+        end
+
+        if errors == {}
+          queries.each_with_index do |query, index|
+            query.run
+            if query.failed
+              errors[index] = query.error_messages
+              statuses.push(500)
+            else
+              results.push(query.results)
+            end
+          end
+        end
+
+        [results, errors, statuses.max]
```

Comment:
yeah.. i know, im not so happy about it
i think it will be better to see if there is a 500 return that or if not you get a 400

Reply:
Actually, I don't think the query runner/validator should be concerned or aware of status code. But that aside, I think the most important is to have actionable error messages when the request is ill-formed (I think it's the case here). Now, to pick between 400 and 500, I would ask myself: can the user do something about it, adapt their query? If yes -> 400, otherwise 500.

Reply:
right, just returning 400 if anything goes wrong, worst case scenario would be that the query runner fails but then you would also have an error message, so its up to the user to figure it out. at least for now.